workers/unity/Assets/Playground/Scripts/Player/MobilePlayerInputSystem.cs

```diff
@@ -0,0 +1,109 @@
+        private static VirtualJoystick cameraJoystick;
+        private static HashSet<int> oldTouchSet = new HashSet<int>();
+        private static HashSet<int> newTouchSet = new HashSet<int>();
+
+        protected override void OnCreateManager(int capacity)
+        {
+            base.OnCreateManager(capacity);
+            try
+            {
+                controllerJoystick = GameObject.FindGameObjectWithTag("MovementJoystick");
+                movementJoystick = controllerJoystick.GetComponent<VirtualJoystick>();
+            }
+            catch (NullReferenceException)
+            {
+                WorkerRegistry.GetWorkerForWorld(World)
+                    .View.LogDispatcher.HandleLog(LogType.Error,
+                        new LogEvent("Could not find movement virtual stick. Movement is now disabled on mobile"));
+            }
+
+            try
+            {
+                cameraControllerJoystick = GameObject.FindGameObjectWithTag("CameraJoystick");
+                cameraJoystick = cameraControllerJoystick.GetComponent<VirtualJoystick>();
+            }
+            catch (NullReferenceException)
+            {
+                WorkerRegistry.GetWorkerForWorld(World).View.LogDispatcher.HandleLog(LogType.Error,
+                    new LogEvent("Could not find virtual camera joystick. Camera movement is now disabled on mobile"));
+            }
+        }
+
+        protected override void OnUpdate()
+        {
+            for (var i = 0; i < inputData.Length; i++)
+            {
+                var input = inputData.PlayerInput[i];
+                var touches = GetNonJoystickTouches();
+                input.Horizontal = movementJoystick.InputDirection.x;
+                input.Vertical = movementJoystick.InputDirection.y;
+                input.RightStickHorizontal = cameraJoystick.InputDirection.x;
+                input.RightStickVertical = cameraJoystick.InputDirection.y;
+                // Camera zoom and running are not implemented for mobile
+                // input.CameraDistance = Input.GetAxis("Mouse ScrollWheel");
+                // input.Running = Input.GetKey(KeyCode.LeftShift);
+                input.ShootSmall = touches == 1;
+                input.ShootLarge = touches >= 2;
+                inputData.PlayerInput[i] = input;
+            }
+        }
+
+
+        private static int GetNonJoystickTouches()
+        {
+            var count = 0;
+            newTouchSet.Clear();
+            foreach (var touch in Input.touches)
+            {
+                if (oldTouchSet.Contains(touch.fingerId))
+                {
+                    newTouchSet.Add(touch.fingerId);
+                }
+            }
+
+            var temp = newTouchSet;
+            newTouchSet = oldTouchSet;
+            oldTouchSet = temp;
+
+            foreach (var touch in Input.touches)
+            {
+                if (!(RectTransformUtility.RectangleContainsScreenPoint(
```

Comment:
Please move the `GetComponent`'s out of the if, restructure this a little bit.

Reply:
Do `!oldTouchSet.Contains(touch.fingerId)` first, as this check is much cheaper.